ortools/linear_solver/xpress_interface.cc

```diff
@@ -1931,9 +1999,104 @@ void XpressInterface::AddSolutionHintToOptimizer() {
     colind[i] = solver_->solution_hint_[i].first->index();
     val[i] = solver_->solution_hint_[i].second;
   }
-  if (int status = XPRSaddmipsol(mLp, len, val.get(), colind.get(), "USER_HINT")) {
-    LOG(WARNING) << "Failed to set solution hint.";
+  XPRSaddhint(mLp, len, val.get(), colind.get(), "USER_HINT");
+}
+
+void XpressInterface::SetCallback(MPCallback* mp_callback) {
+  if (callback_ != nullptr) {
+    // replace existing callback by removing it first
+    CHECK_STATUS(XPRSremovecboptnode(mLp, XpressOptNodeCallbackImpl, NULL));
+  }
+  callback_ = mp_callback;
+}
+
+// NOTE(user): This function must have this exact API, because we are passing
+// it to XPRESS as a callback.
+void XPRS_CC XpressOptNodeCallbackImpl(XPRSprob cbprob, void* cbdata, int* p_infeasible) {
+  auto callback_with_context = static_cast<const MPCallbackWithXpressContext*>(cbdata);
+  CHECK(callback_with_context != nullptr);
+  CHECK(callback_with_context->context != nullptr);
+  CHECK(callback_with_context->callback != nullptr);
+  if (callback_with_context->context->UpdateFromXpressState(cbprob)) {
+    // Only run user's callback if the context has changed
+    callback_with_context->callback->RunCallback(
+        callback_with_context->context);
+  }
+}
+
+XpressMPCallbackContext::XpressMPCallbackContext(XPRSprob& mLp, bool mip,
+                                                 bool maximize, int num_vars)
+    : mLp_(mLp),
+      mip_(mip),
+      maximize_(maximize),
+      num_vars_(num_vars),
+      variable_values_(num_vars_, XPRS_NAN),
+      objective_value_(XPRS_NAN),
+      best_objective_bound_(XPRS_NAN),
+      num_nodes_(0),
+      num_user_sols_(0) {}
+
+MPCallbackEvent XpressMPCallbackContext::Event() {
+  return MPCallbackEvent::kMipSolution; // this is the only supported callback for now
+}
+
+bool XpressMPCallbackContext::CanQueryVariableValues() {
+  return !std::isnan(objective_value_);
+}
+
+double XpressMPCallbackContext::VariableValue(const MPVariable* variable) {
+  return variable_values_[variable->index()];
+}
+
+double XpressMPCallbackContext::SuggestSolution(
+    const absl::flat_hash_map<const MPVariable*, double>& solution) {
+  // Currently the XPRESS API does not handle clearing out previous hints
+  const std::size_t len = solution.size();
+  if (len == 0) {
+    // hint is empty, do nothing
+    return XPRS_NAN;
+  }
+  unique_ptr<int[]> colind(new int[len]);
+  unique_ptr<double[]> val(new double[len]);
+  int i = 0;
+  for (const auto& [var, value] : solution) {
+    colind[i] = var->index();
+    val[i] = value;
+    ++i;
   }
+  // use a new name so new solutions don't override old ones
+  std::string name = "USER_CB_HINT_" + std::to_string(++num_user_sols_);
+  XPRSaddhint(mLp_, len, val.get(), colind.get(), name.c_str());
```

Comment:
Passing `mLP_` as first argument is wrong here. You have to pass the `XPRSprob` pointer that was passed into function `XpressOptNodeCallbackImpl`. This is different on different threads and almost always different from `mLP_`.

Reply:
I think this would not be an issue since multithreading would require the user to initialize multiple linear solver instances, thus creating different mLP_ instances.
However, I do get your point and I will change the code like you suggested. It does eliminate any risk, plus it makes the code cleaner :)